README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

nebula
======

<!-- badges: start -->

[![Lifecycle:
maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
<!-- [![CRAN status](https://www.r-pkg.org/badges/version/nebula)](https://CRAN.R-project.org/package=nebula) -->
[![Codecov test
coverage](https://codecov.io/gh/nebula-group/nebula/branch/master/graph/badge.svg)](https://codecov.io/gh/nebula-group/nebula?branch=master)
[![R-CMD-CHECK {via
tic}](https://github.com/margarethannum/nebula/workflows/tic/badge.svg?branch=master)](https://github.com/margarethannum/nebula/actions)
[![R build
status](https://github.com/nebula-group/nebula/workflows/R-CMD-check/badge.svg)](https://github.com/nebula-group/nebula/actions)
<!-- badges: end -->

The Nebula package implements Network-based latent-Dirichlet subtype
analysis (Nebula) algorithm.

Practically, this can be used to incorporate biological
networks/pathways to inform clustering solutions.

Flexible sparsity parameters for multiple input data types allows for
control over which data types need sparse vs rich feature selection.

Installation
------------

You can install the development version from
[GitHub](https://github.com/nebula-group/nebula) with:

    # install.packages("remotes")

    remotes::install_github("nebula-group/nebula", build_vignettes = TRUE)

    library(nebula)

Learn more in `vignette("nebula_tutorial", package = "nebula")` or
`?nebula`.


NEWS.md
# nebula (development version)

* Added widespread input checks and unit testing
* changed package name from Nebula --> nebula
* Incorporated Continuous in

# Nebula 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(nebula)

test_check("nebula")


tests/testthat/test-Nebula.R
context("data and parameter inputs")

#TODO test for binary data being of 2 types
#TODO test for continuous input numeric
#TODO test for E input - four columns and appropriate entries

set.seed(123)

#make smaller data for testing purposes
data = lapply(colon$modal, function(x) x[1:100,])

test_that("M modalities have same number of samples", {
  #make vector of number of rows in each matrix M of input data
  colon_bad_modality <- colon$modal
  colon_bad_modality[[1]] <- colon_bad_modality[[1]][1:330,]
  #should error if there are not
  expect_error(nebula(
    data = colon_bad_modality,
    modtype = c(0, 1), #
    E = colon$network,
    H = 3,
    modeta = c(1, 0.2),
    nu = 1,
    alpha = 1,
    lam = 1,
    alpha_sigma = 10,
    beta_sigma = 10,
    alpha_p = 1,
    beta_p = 1,
  ))
})

test_that("Check for missing data", {
  #make vector of number of rows in each matrix M of input data
  colon_missing <- colon$modal
  colon_missing[[2]][1,1] <- NA
  #should error if there are not
  expect_error(nebula(
    data = colon_missing,
    modtype = c(0, 1), #
    E = colon$network,
    H = 3,
    modeta = c(1, 0.2),
    nu = 1,
    alpha = 1,
    lam = 1,
    alpha_sigma = 10,
    beta_sigma = 10,
    alpha_p = 1,
    beta_p = 1,
  ))
})

test_that("modtype incorrect", {
  #make vector of number of rows in each matrix M of input data
  #should error if there are not the right number
  expect_error(nebula(
    data = data,
    modtype = c(0), #
    E = colon$network,
    H = 3,
    modeta = c(1, 0.2),
    nu = 1,
    alpha = 1,
    lam = 1,
    alpha_sigma = 10,
    beta_sigma = 10,
    alpha_p = 1,
    beta_p = 1,
  ))
  #or if modtype is missing
  expect_error(nebula(
    data = data,
   # modtype = c(0), #
    E = colon$network,
    H = 3,
    modeta = c(1, 0.2),
    nu = 1,
    alpha = 1,
    lam = 1,
    alpha_sigma = 10,
    beta_sigma = 10,
    alpha_p = 1,
    beta_p = 1,
  ))
  expect_error(nebula(
    data = data,
    # modtype = c(0), #
    E = colon$network,
    H = 3,
    modeta = c(1, 0.2),
    nu = 1,
    alpha = 1,
    lam = 1,
    alpha_sigma = 10,
    beta_sigma = 10,
    alpha_p = 1,
    beta_p = 1,
  ))
  #or incorrect entry
  expect_error(nebula(
    data = data,
    modtype = c("continuous", "binary"), #
    E = colon$network,
    H = 3,
    modeta = c(1, 0.2),
    nu = 1,
    alpha = 1,
    lam = 1,
    alpha_sigma = 10,
    beta_sigma = 10,
    alpha_p = 1,
    beta_p = 1,
  ))
})

test_that("modeta incorrect", {
  #make vector of number of rows in each matrix M of input data
  #should error if there are not
  expect_error(nebula(
    data = data,
    modtype = c(0, 1), #
    E = colon$network,
    H = 3,
    modeta = c(1), # wrong number of parameters
    nu = 1,
    alpha = 1,
    lam = 1,
    alpha_sigma = 10,
    beta_sigma = 10,
    alpha_p = 1,
    beta_p = 1,
  ))
  # or missing
  expect_error(nebula(
    data = data,
    modtype = c(0, 1), #
    E = colon$network,
    H = 3,
   # modeta = c(1),# missing modet
    nu = 1,
    alpha = 1,
    lam = 1,
    alpha_sigma = 10,
    beta_sigma = 10,
    alpha_p = 1,
    beta_p = 1,
  ))
})

test_that("H input", {
  # #should error if there are not correct input
  # expect_error(nebula(
  #   data = data,
  #   modtype = c(0, 1), #
  #   E = colon$network,
  #   H = 3,
  #   modeta = c(0,1),
  #   nu = 1,
  #   alpha = 1,
  #   lam = 1,
  #   alpha_sigma = 10,
  #   beta_sigma = 10,
  #   alpha_p = 1,
  #   beta_p = 1,
  # ))
  # or missing
  expect_error(nebula(
    data = data,
    modtype = c(0, 1), #
    E = colon$network,
    #H = 3, #missing
    modeta = c(0, 1),
    nu = 1,
    alpha = 1,
    lam = 1,
    alpha_sigma = 10,
    beta_sigma = 10,
    alpha_p = 1,
    beta_p = 1,
  ))
})

test_that("nu input", {
  # #should error if there are not correct input
  # expect_error(nebula(
  #   data = data,
  #   modtype = c(0, 1), #
  #   E = colon$network,
  #   H = 3,
  #   modeta = c(0,1),
  #   nu = 1,
  #   alpha = 1,
  #   lam = 1,
  #   alpha_sigma = 10,
  #   beta_sigma = 10,
  #   alpha_p = 1,
  #   beta_p = 1,
  # ))
  # or missing
  expect_error(nebula(
    data = data,
    modtype = c(0, 1), #
    E = colon$network,
    H = 3,
    modeta = c(0, 1),
    #nu = 1, #missing
    alpha = 1,
    lam = 1,
    alpha_sigma = 10,
    beta_sigma = 10,
    alpha_p = 1,
    beta_p = 1,
  ))
})

test_that("alpha input", {
  # #should error if there are not correct input
  # expect_error(nebula(
  #   data = data,
  #   modtype = c(0, 1), #
  #   E = colon$network,
  #   H = 3,
  #   modeta = c(0,1),
  #   nu = 1,
  #   alpha = 1,
  #   lam = 1,
  #   alpha_sigma = 10,
  #   beta_sigma = 10,
  #   alpha_p = 1,
  #   beta_p = 1,
  # ))
  # or missing
  expect_error(nebula(
    data = data,
    modtype = c(0, 1), #
    E = colon$network,
    H = 3,
    modeta = c(0, 1),
    nu = 1, #missing
    #alpha = 1,
    lam = 1,
    alpha_sigma = 10,
    beta_sigma = 10,
    alpha_p = 1,
    beta_p = 1,
  ))
})

test_that("lam input", {
  # #should error if there are not correct input
  # expect_error(nebula(
  #   data = data,
  #   modtype = c(0, 1), #
  #   E = colon$network,
  #   H = 3,
  #   modeta = c(0,1),
  #   nu = 1,
  #   alpha = 1,
  #   lam = 1,
  #   alpha_sigma = 10,
  #   beta_sigma = 10,
  #   alpha_p = 1,
  #   beta_p = 1,
  # ))
  # or missing
  expect_error(nebula(
    data = data,
    modtype = c(0, 1), #
    E = colon$network,
    H = 3,
    modeta = c(0, 1),
    nu = 1, #missing
    alpha = 1,
    #lam = 1,
    alpha_sigma = 10,
    beta_sigma = 10,
    alpha_p = 1,
    beta_p = 1,
  ))
})

test_that("pr0 input", {
  # #should error if there are not correct input
  expect_error(nebula(
    data = data,
    modtype = c(0, 1), #
    E = colon$network,
    H = 3,
    modeta = c(0,1),
    nu = 1,
    alpha = 1,
    lam = 1,
    alpha_sigma = 10,
    beta_sigma = 10,
    alpha_p = 1,
    beta_p = 1,
    pr0 = -1
  ))
  # or missing
  expect_error(nebula(
    data = data,
    modtype = c(0, 1), #
    E = colon$network,
    H = 3,
    modeta = c(0, 1),
    nu = 1, #missing
    alpha = 1,
    lam = 1,
    alpha_sigma = 10,
    beta_sigma = 10,
    alpha_p = 1,
    beta_p = 1,
    pr0 = 2
  ))
})

test_that("sig0 input", {
  # #should error if there are not correct input
  expect_error(nebula(
    data = data,
    modtype = c(0, 1), #
    E = colon$network,
    H = 3,
    modeta = c(0,1),
    nu = 1,
    alpha = 1,
    lam = 1,
    alpha_sigma = 10,
    beta_sigma = 10,
    alpha_p = 1,
    beta_p = 1,
    sig0 = -1
  ))
})

test_that("binit input", {
  n_bad <- 99
  H <- 3
  binit_bad <- matrix(stats::rnorm(n_bad*H,0,0.01),n_bad,H)
  # #should error if there are not correct input
  expect_error(nebula(
    data = data,
    modtype = c(0, 1), #
    E = colon$network,
    H = 3,
    modeta = c(0,1),
    nu = 1,
    alpha = 1,
    lam = 1,
    alpha_sigma = 10,
    beta_sigma = 10,
    alpha_p = 1,
    beta_p = 1,
    binit = binit_bad
  ))
})

test_that("E input", {
  e_bad <- colon$network[,1:3]
  # #should error if there are not correct input
  expect_error(nebula(
    data = data,
    modtype = c(0, 1), #
    E = e_bad,
    H = 3,
    modeta = c(0,1),
    nu = 1,
    alpha = 1,
    lam = 1,
    alpha_sigma = 10,
    beta_sigma = 10,
    alpha_p = 1,
    beta_p = 1
  ))
})



tests/testthat/test-NebulaCore.R
context("inner workings of core formula")

library(nebula)

set.seed(123)

#set all data as is done in shell nebula function
#make smaller data for testing purposes
data = lapply(colon$modal, function(x) x[1:100,])


modtype = c(0, 1)
E = colon$network
H = 3
modeta = c(1, 0.2)
nu = 1
alpha = 1
lam = 1
alpha_sigma = 10
beta_sigma = 10
alpha_p = 1
beta_p = 1

# now do some prep to the data as in nebula() function, get ready
# to pass to NebulaCore()
M <- length(data)

p <- rep(0, M)

for (m in 1:M)
{
  if (m == 1) {
    n <- nrow(data[[1]])
  } else if (nrow(data[[m]]) != n) {
    stop("data matrices need to have same number of samples (rows)")
  }
  p[m] <- ncol(data[[m]])
}

P <- sum(p)
cump <- stats::diffinv(p)

n <- nrow(data[[1]])

X <- matrix(0, n, P)
mod <- matrix(FALSE, P, M) # modality
type <- rep(0, P) # variable type

E2 <- E[, c(2, 4)] + cump[E[, c(1, 3)]]
eta <- rep(0, P) # sparsity parameter for each variable
for (m in 1:M)
{
  X[, (cump[m] + 1):cump[m + 1]] <- data[[m]]
  mod[(cump[m] + 1):cump[m + 1], m] <- TRUE
  type[(cump[m] + 1):cump[m + 1]] <- modtype[m]
  eta[(cump[m] + 1):cump[m + 1]] <- modeta[m]
}

# set other parameters
alpha_sigma = 1
beta_sigma = 1
alpha_p = 1
beta_p = 1
mu0 = 0
sig0 = 20
pr0 = 0.5
binit = NULL

out <- nebula:::NebulaCore(X, type, E2, H, eta, nu, alpha, lam, alpha_sigma, beta_sigma, alpha_p, beta_p, mu0, sig0, pr0, binit = NULL)

test_that("multiplication works", {
  expect_equal(2 * 2, 4)
})


R/DMMVI.R
#' DMMVI
#'
#' Variational Inference for Dirichlet Mixture Model
#' (ver. 20191126)
#'
#' @param X n by p matrix where n is the sample size and p is the feature size
#' @param type p-dimensional vector of feature types. Currently supports continuous(=0) and binary(=1)
#' @param A the adjacency matrix of the graph representing the graphical structure of X
#' @param H the number of clusters to be fit
#' @param eta sparsity parameter for gamma's
#' @param nu smoothness parameter for gamma's
#' @param alpha concentration parameter for dirichlet process
#' @param lam NEED DESCRIPTION
#' @param mu0 mean of the non-selected continuous features
#' @param sig0 variance of the non-selected continuous features
#' @param pr0 "on" probability of the non-selected binary features
#' @param alpha_sigma shape parameter of the prior of residual variance(sigma^2)
#' @param beta_sigma rate parameter of the prior of residual variance(sigma^2)
#' @param alpha_p first shape parameter of the prior of the "on" probabilities(p_hj) of binary features
#' @param beta_p second shape parameter of the prior of the "on" probabilities(p_hj) of binary features
#' @param binit n by H initial matrix of B, exp(B_ih) is proportional to Pr(z_i=h). If NULL (default), random numbers are filled in.
#' @export
#' @author Changgee Chang
#' @examples
#' # ADD EXAMPLE
#
DMMVI <- function(X,type,A,H,eta,nu,alpha,lam,mu0,sig0,pr0,alpha_sigma,beta_sigma,alpha_p,beta_p,binit=NULL)
{
  n = nrow(X)
  p = ncol(X)
  idx0 = which(type==0)
  idx1 = which(type==1)
  p0 = length(idx0)
  p1 = length(idx1)

  # Initialize
  if ( is.null(binit) )
    b = matrix(stats::rnorm(n*H,0,0.01),n,H)
  else
    b = binit
  b = b - apply(b,1,max)
  eb = exp(b)
  EI = eb/apply(eb,1,sum)

  c = matrix(0,p,H)
  cbase = matrix(0,p,H)
  Egam = 1/(1+exp(-c))

  Elgau = array(0,c(n,p,H))
  Elbin = array(0,c(n,p,H))
  lgau0 = (X[,idx0] - mu0)^2/sig0 + log(sig0)
  lbin0 = X[,idx1]*log(pr0) + (1-X[,idx1])*log(1-pr0)

  iter = 0
  while (TRUE)
  {
    iter = iter + 1

    # Step 1
    sEI = apply(EI,2,sum)
    f = 1 + sEI
    tmp = stats::diffinv(-sEI)
    g = alpha + tmp[-1] - tmp[H+1]

    tmp = digamma(f+g)
    Elw = digamma(f) - tmp
    El1w = digamma(g) - tmp
    Elw[H] = 0
    El1w[H] = -Inf
    cumEl1w = stats::diffinv(El1w)


    # Step 2
    EgamsEI = Egam*rep(sEI,each=p)
    XEI = t(X)%*%EI
    EgamXEI = Egam*XEI
    d = alpha_sigma + EgamsEI
    v = lam + EgamsEI
    m = EgamXEI/v
    r = beta_sigma - v*m^2 + Egam*(t(X^2)%*%EI)

    Elsig = log(r/2) - digamma(d/2)
    for ( h in 1:H )
      Elgau[,,h] = t((t(X) - m[,h])^2*d[,h]/r[,h] + 1/v[,h] + Elsig[,h])


    # Step 3
    s = alpha_p + EgamXEI
    t = beta_p + EgamsEI - EgamXEI
    s[idx0,] = 1
    t[idx0,] = 1

    tmp = digamma(s+t)
    Elp = digamma(s) - tmp
    El1p = digamma(t) - tmp
    for ( h in 1:H )
      Elbin[,,h] = t(t(X)*Elp[,h]+(1-t(X))*El1p[,h])


    # Step 4
    for ( h in 1:H )
    {
      cbase[idx0,h] = - t(Elgau[,idx0,h]) %*% EI[,h]/2
      cbase[idx0,h] = cbase[idx0,h] + t(lgau0) %*% EI[,h]/2
      cbase[idx1,h] = t(Elbin[,idx1,h]) %*% EI[,h]
      cbase[idx1,h] = cbase[idx1,h] - t(lbin0) %*% EI[,h]
    }
    cbase = cbase - eta

    while (TRUE)
    {
      pc = c
      for ( j in 1:p )
      {
        c[j,] = cbase[j,] + nu*A[j,]%*%(2*Egam-1)
        Egam[j,] = 1/(1+exp(-c[j,]))
      }
      if ( max(abs(pc-c)) < 1e-4 )
        break
    }


    # Step 5
    pb = b
    for ( h in 1:H )
    {
      b[,h] = - Elgau[,idx0,h] %*% Egam[idx0,h]/2
      b[,h] = b[,h] - lgau0 %*% (1-Egam[idx0,h])/2
      b[,h] = b[,h] + Elbin[,idx1,h] %*% Egam[idx1,h]
      b[,h] = b[,h] + lbin0 %*% (1-Egam[idx1,h])
      b[,h] = b[,h] + Elw[h] + cumEl1w[h]
    }
    b = b - apply(b,1,max)
    eb = exp(b)
    EI = eb/apply(eb,1,sum)

    if ( max(abs(pb-b)) < 1e-4 )
      break
  }

  list(EI=EI,Egam=Egam,m=m,iter=iter)
}



R/NebulaCore.R
#' NebulaCore
#'
#' Network based latent dirichlet subtype analysis
#' (ver. 20190920)
#'
#' @param X n by p matrix where n is the sample size and p is the feature size
#' @param type p-dimensional vector of feature types. Currently supports continuous(=0) and binary(=1)
#' @param E e by 2 matrix of the graph edges representing the graphical structure of X
#' @param H the number of clusters to be fit
#' @param eta length p vector of sparsity parameters
#' @param nu smoothness parameter for gamma's
#' @param alpha concentration parameter for dirichlet process
#' @param lam shrinkage parameter for means of selected continuous features
#' @param alpha_sigma shape parameter of the prior of residual variance(sigma^2)
#' @param beta_sigma rate parameter of the prior of residual variance(sigma^2)
#' @param alpha_p first shape parameter of the prior of the 'active' probabilities(p_hj) of binary features
#' @param beta_p second shape parameter of the prior of the 'active' probabilities(p_hj) of binary features
#' @param mu0 mean of the non-selected continuous features
#' @param sig0 variance of the non-selected continuous features
#' @param pr0 'active' probability of the non-selected binary features
#' @param binit n by H initial matrix of B, exp(B_ih) is proportional to Pr(z_i=h). If NULL (default), random numbers are filled in.
#' @noRd
#' @keywords internal
#' @author Changgee Chang
#' @examples
#' # ADD EXAMPLE

NebulaCore <- function(X,type,E,H,eta,nu,alpha,lam,alpha_sigma,beta_sigma,alpha_p,beta_p,mu0,sig0,pr0,binit=NULL)
{
  n = nrow(X)
  p = ncol(X)
  idx0 = which(type==0)
  idx1 = which(type==1)
  p0 = length(idx0)
  p1 = length(idx1)

  E = E[order(E[,1],E[,2]),]
  e = nrow(E)
  nadj = rep(0,p)
  tmp = c(diff(E[,1])!=0,TRUE)
  nadj[E[tmp,1]] = diff(c(0,which(tmp)))
  Eidx = stats::diffinv(nadj)

  # Initialize
  if ( is.null(binit) )
    b = matrix(stats::rnorm(n*H,0,0.01),n,H)
  else
    b = binit
  b = b - apply(b,1,max)
  eb = exp(b)
  EI = eb/apply(eb,1,sum)

  c = matrix(0,p,H)
  cbase = matrix(0,p,H)
  Egam = 1/(1+exp(-c))

  Elgau = array(0,c(n,p,H))
  Elbin = array(0,c(n,p,H))
  lgau0 = (X[,idx0] - mu0)^2/sig0 + log(sig0)
  lbin0 = X[,idx1]*log(pr0) + (1-X[,idx1])*log(1-pr0)

  iter = 0
  while (TRUE)
  {
    iter = iter + 1

    # Step 1
    sEI = apply(EI,2,sum)
    f = 1 + sEI
    tmp = stats::diffinv(-sEI)
    g = alpha + tmp[-1] - tmp[H+1]

    tmp = digamma(f+g)
    Elw = digamma(f) - tmp
    El1w = digamma(g) - tmp
    Elw[H] = 0
    El1w[H] = -Inf
    cumEl1w = stats::diffinv(El1w)


    # Step 2
    EgamsEI = Egam*rep(sEI,each=p)
    XEI = t(X)%*%EI
    EgamXEI = Egam*XEI
    d = alpha_sigma + EgamsEI
    v = lam + EgamsEI
    m = EgamXEI/v
    r = beta_sigma - v*m^2 + Egam*(t(X^2)%*%EI)

    Elsig = log(r/2) - digamma(d/2)
    for ( h in 1:H )
      Elgau[,,h] = t((t(X) - m[,h])^2*d[,h]/r[,h] + 1/v[,h] + Elsig[,h])


    # Step 3
    s = alpha_p + EgamXEI
    t = beta_p + EgamsEI - EgamXEI
    s[idx0,] = 1
    t[idx0,] = 1

    tmp = digamma(s+t)
    Elp = digamma(s) - tmp
    El1p = digamma(t) - tmp
    for ( h in 1:H )
      Elbin[,,h] = t(t(X)*Elp[,h]+(1-t(X))*El1p[,h])


    # Step 4
    for ( h in 1:H )
    {
      cbase[idx0,h] = - t(Elgau[,idx0,h]) %*% EI[,h]/2
      cbase[idx0,h] = cbase[idx0,h] + t(lgau0) %*% EI[,h]/2 - eta[idx0]
      cbase[idx1,h] = t(Elbin[,idx1,h]) %*% EI[,h]
      cbase[idx1,h] = cbase[idx1,h] - t(lbin0) %*% EI[,h] - eta[idx1]
    }

    Egam[nadj==0,] = 1/(1+exp(-cbase[nadj==0,]))
    while (TRUE)
    {
      pc = c
      for ( j in which(nadj!=0) )
      {
        if ( nadj[j] == 1 )
          c[j,] = cbase[j,] + nu*(2*Egam[E[Eidx[j+1],2],]-1)
        else
          c[j,] = cbase[j,] + nu*apply(2*Egam[E[(Eidx[j]+1):Eidx[j+1],2],]-1,2,sum)
        Egam[j,] = 1/(1+exp(-c[j,]))
      }
      if ( max(abs(pc-c)) < 1e-4 )
        break
    }


    # Step 5
    pb = b
    for ( h in 1:H )
    {
      b[,h] = - Elgau[,idx0,h] %*% Egam[idx0,h]/2
      b[,h] = b[,h] - lgau0 %*% (1-Egam[idx0,h])/2
      b[,h] = b[,h] + Elbin[,idx1,h] %*% Egam[idx1,h]
      b[,h] = b[,h] + lbin0 %*% (1-Egam[idx1,h])
      b[,h] = b[,h] + Elw[h] + cumEl1w[h]
    }
    b = b - apply(b,1,max)
    eb = exp(b)
    EI = eb/apply(eb,1,sum)

    if ( max(abs(pb-b)) < 1e-4 )
      break
  }

  list(EI=EI,Egam=Egam,m=m,lpr=Elp,iter=iter)
}



R/Nebula.R
#' nebula
#'
#' Network based latent dirichlet subtype analysis
#' (ver. 20190920)
#'
#' @param data list of M data matrices, where each matrix is n samples by p_m features for modality M
#' @param modtype M-length vector of feature types for M modalities. Currently supports continuous(=0) and binary(=1)
#' @param E e by 4 matrix, each tuple (row) of which represents an edge; (m1,j1,m2,j2) variable j1 of modality m1 is connected to variable j2 of modality m2
#' @param H the number of clusters to be fit
#' @param modeta length M vector of sparsity parameters for M modalities
#' @param nu smoothness parameter for gamma's
#' @param alpha concentration parameter for dirichlet process
#' @param lam shrinkage parameter for means of selected continuous features
#' @param alpha_sigma shape parameter of the prior of residual variance(sigma^2) (default = 1, i.e. noninformative)
#' @param beta_sigma rate parameter of the prior of residual variance(sigma^2) (default = 1, i.e. noninformative)
#' @param alpha_p first shape parameter of the prior of the 'active' probabilities(p_hj) of binary features (default = 1, i.e. noninformative)
#' @param beta_p second shape parameter of the prior of the 'active' probabilities(p_hj) of binary features (default = 1, i.e. noninformative)
#' @param mu0 mean of the non-selected continuous features (default is 0)
#' @param sig0 variance of the non-selected continuous features (default is 20)
#' @param pr0 'active' probability of the non-selected binary features (default is 0.5)
#' @param binit n by H initial matrix of B, exp(B_ih) is proportional to Pr(z_i=h). If NULL (default), random numbers are filled in.
#'
#' @return A list containing clustering assignments, variable selection, and
#'   posterior probabilities
#' \itemize{
#' \item clustering cluster assignment
#' \item defvar list of M matrices; each matrix is p_m by H indicating the variable j in modality m is a defining variable for the cluster h.
#' \item clus_pr n by H matrix containing the probability that the subject i belongs to the cluster h.
#' \item defvar_pr list of M matrices; each matrix is p_m by H containing the probability that the variable j in modality m is a defining variable for the cluster h.
#' \item def_m list of M matrices; each matrix is p_m by H containing the mean of the variable j as a defining variable for the cluster h. (continuous variable only)
#' \item def_lpr list of M matrices: each matrix is p_m by H containing the log probabilities of the variable j being 'active' and a defining variable for the cluster h. (binary variable only)
#' \item iter the number of iterations until the algorithm converges.
#' \item param A list of the input parameters used in the clustering solution.
#' }
#' @export
#' @author Changgee Chang
#' @examples
#'
#' \dontrun{
#' res <- nebula(
#'          data = colon$modal,
#'          modtype = c(0, 1),
#'          E = colon$network,
#'          H = 3,
#'          modeta = c(1, 0.2),
#'          nu = 1,
#'          alpha = 1,
#'          lam = 1,
#'          alpha_sigma = 10,
#'          beta_sigma = 10,
#'          alpha_p = 1,
#'          beta_p = 1,
#'          )
#'  }
#'
nebula <- function(data, modtype, E, H, modeta, nu, alpha, lam, alpha_sigma = 1,
                   beta_sigma = 1, alpha_p = 1, beta_p = 1, mu0 = 0, sig0 = 20, pr0 = 0.5, binit = NULL) {

# check inputs for missing or incorrect
  if(missing(modtype))
    stop("Must specify model type(s) for data. modtype supports continuous (=0) or binary (=1).")
  if(missing(modeta))
    stop("Must specify modeta sparsity parameters for M modalities.")
  if(missing(H))
    stop("Must specify H for the number of clusters to be fit.")
  if(missing(nu))
    stop("Must specify nu for smoothness parameter for gammas.")
  if(missing(alpha))
    stop("Must specify alpha concentration parameter for dirichlet process.")
  if(missing(lam))
    stop("Must specify lam shrinkage parameter for selected continuous features.")
  if(pr0>1 | pr0 <0)
    stop("pr0 must be a probability between 0 and 1.")
  if(sig0 <0)
    stop("sig0 must be a positive value to define variance of the non-selected continuous features.")

  M <- length(data)

  if (length(modtype) != M) {
    stop("length of modtype needs to match number of matrices in input data")
  }
  if (length(modeta) != M) {
    stop("length of modeta needs to match number of matrices in input data")
  }

  p <- rep(0, M)
  #checks for same number of samples
  for (m in 1:M)
  {
    if (m == 1) {
      n <- nrow(data[[1]])
    } else if (nrow(data[[m]]) != n) {
      stop("Input data matrices need to have same number of samples (rows)")
    }
    p[m] <- ncol(data[[m]])
  }

  #check for missing data
  if(sum(!stats::complete.cases(data))>0)
    stop("Data cannot have NAs. Please exclude or impute missing values.")

  #check that binit is correct dimension
  if ( !is.null(binit) )
    if(dim(binit)[1] != n | dim(binit)[2] != H)
      stop("binit must be NULL or have dimensions n by H.")


  P <- sum(p) #sum of features across all modalities
  cump <- stats::diffinv(p)

  X <- matrix(0, n, P) #blank matrix of n samples by P total features
  mod <- matrix(FALSE, P, M) # modality
  type <- rep(0, P) # variable type
  if (ncol(E) != 4) {
    stop("Matrix E must have 4 columns")
  }
  E2 <- E[, c(2, 4)] + cump[E[, c(1, 3)]]
  eta <- rep(0, P) # sparsity parameter for each variable
  for (m in 1:M)
  {
    #make single data matrix for input into core nebula formula
    X[, (cump[m] + 1):cump[m + 1]] <- data[[m]]
    mod[(cump[m] + 1):cump[m + 1], m] <- TRUE
    if (modtype[m] != 0 & modtype[m] != 1) {
      stop("modtype must be specified as 0 (continuous) or 1 (binary) for each modality")
    }
    type[(cump[m] + 1):cump[m + 1]] <- modtype[m]
    eta[(cump[m] + 1):cump[m + 1]] <- modeta[m]
  }

  out <- NebulaCore(X, type, E2, H, eta, nu, alpha, lam, alpha_sigma, beta_sigma, alpha_p, beta_p, mu0, sig0, pr0, binit)

  defvar <- defvar_pr <- def_m <- def_lpr <- list()

  for (m in 1:M)
  {
    defvar[[m]] <- matrix(out$Egam[mod[, m], ] > 0.5, ncol = H)
    defvar_pr[[m]] <- matrix(out$Egam[mod[, m], ], ncol = H)
    if (modtype[m] != 0) { out$m[mod[, m], ] <- NA}
    def_m[[m]] <- matrix(out$m[mod[, m], ], ncol = H)
    if (modtype[m] != 1) {out$lpr[mod[, m], ] <- NA}
    def_lpr[[m]] <- matrix(out$lpr[mod[, m], ], ncol = H)
  }

  param <- list(modtype = modtype, E = E, H = H, modeta = modeta, nu = nu, alpha = alpha, lam = lam, alpha_sigma = alpha_sigma, beta_sigma = beta_sigma, alpha_p = alpha_p, beta_p = beta_p, mu0 = mu0, sig0 = sig0, pr0 = pr0, binit = binit)

  # Outputs
  #
  # clustering: cluster assignment
  # defvar: list of M matrices; each matrix is p_m by H indicating the variable j in modality m is a defining variable for the cluster h.
  # clus_pr: n by H matrix containing the probability that the subject i belongs to the cluster h.
  # defvar_pr: list of M matrices; each matrix is p_m by H containing the probability that the variable j in modality m is a defining variable for the cluster h.
  # def_m: list of M matrices; each matrix is p_m by H containing the mean of the variable j as a defining variable for the cluster h. (continuous variable only)
  # def_lpr: list of M matrices: each matrix is p_m by H containing the log probabilities of the variable j being 'active' ad a defining variable for the cluster h. (binary variable only)
  # iter: the number of iterations until the algorithm converges.
  list(clustering = apply(out$EI, 1, which.max), defvar = defvar, clus_pr = out$EI, defvar_pr = defvar_pr, def_m = def_m, def_lpr = def_lpr, iter = out$iter, param = param)
}
